shardmanager.go

```diff
@@ -90,67 +90,86 @@ func (n *ShardManager) connectShardNodes(ctx context.Context, shardID ShardIDTyp
 	}
 
 	// borrowed from `bootstrapConnect`, should be modified/refactored and tested
-	errs := make(chan error, len(pinfos))
 	var wg sync.WaitGroup
 	for _, p := range pinfos {
 		wg.Add(1)
-		go func(p pstore.PeerInfo) {
+		go func(ctx context.Context, p pstore.PeerInfo) {
+			// Add span for Connect of ShardManager.connectShardNodes
+			childSpanctx := logger.Start(ctx, "ShardManager.connectShardNodes.Connect")
+			defer logger.Finish(childSpanctx)
+
 			defer wg.Done()
 			if err := n.host.Connect(ctx, p); err != nil {
+				logger.SetErr(childSpanctx, fmt.Errorf("Failed to connect peer %v in shard %v, err: %v", p.ID, shardID, err))
 				log.Printf(
 					"Failed to connect peer %v in shard %v: %s",
 					p.ID,
 					shardID,
 					err,
 				)
-				errs <- err
 				return
 			}
 			log.Printf("Successfully connected peer %v in shard %v", p.ID, shardID)
-		}(p)
+			logger.SetTag(childSpanctx, "peer", p.ID)
+		}(spanctx, p)
 	}
 	wg.Wait()
 	// FIXME: ignore the errors when connecting shard peers for now
 	return nil
 }
 
-// TODO: return error if it fails
-func (n *ShardManager) ListenShard(ctx context.Context, shardID ShardIDType) {
-	// Add span for AddPeer of ShardManager
-	span, newctx := opentracing.StartSpanFromContext(ctx, "ShardManager.ListenShard")
-	defer span.Finish()
-	// Set shardID info in Baggage
-	span.SetBaggageItem("shardID", fmt.Sprintf("%v", shardID))
+func (n *ShardManager) ListenShard(ctx context.Context, shardID ShardIDType) error {
+	// Add span for ListenShard of ShardManager
+	spanctx := logger.Start(ctx, "ShardManager.ListenShard")
+	defer logger.Finish(spanctx)
 
 	if n.IsShardListened(shardID) {
-		return
+		return nil
 	}
 	n.shardPrefTable.AddPeerListeningShard(n.host.ID(), shardID)
 
 	// TODO: should set a critiria: if we have enough peers in the shard, don't connect shard nodes
-	n.connectShardNodes(newctx, shardID)
+	if err := n.connectShardNodes(spanctx, shardID); err != nil {
+		logger.FinishWithErr(spanctx, fmt.Errorf("Failed to connect to nodes in shard %v", shardID))
+		log.Printf("Failed to connect to nodes in shard %v", shardID)
+		return err
```

Comment:
I don't think we should `return` here. Even though we fail to connect to the "shard peers", we should keep doing `n.SubscribeCollation`.